### PR TITLE
Fixed #27818 -- Used contextib.suppress for suppressing exceptions.

### PR DESCRIPTION
**Pull request by atombrella**
Not complete, work in progress. There are quite a lot in the pattern which are (naturally ignored).
```
try: ... : 
except: 
   pass
else:
   ... 
```

### CODE REVIEW

**Comment by atombrella** [comment]
I specifically didn't touch `six.py`; I think it'll be removed at some point anyway.

`sqlite3/schema.py` has a pattern that can't be replaced, although both exceptions are silenced (`import  sqlite3` is required).

File: django/contrib/gis/gdal/srs.py [hunk 1]
```diff
@@ -55,13 +56,11 @@ def __init__(self, srs_input='', srs_type='user'):
             self.import_wkt(srs_input)
             return
         elif isinstance(srs_input, str):
-            try:
+            with suppress(ValueError):
                 # If SRID is a string, e.g., '4326', then make acceptable
                 # as user input.
                 srid = int(srs_input)
                 srs_input = 'EPSG:%d' % srid
```
Comment (timgraham) [inline django/contrib/gis/gdal/srs.py]: I think this line shouldn't be in the context manager.

File: django/contrib/gis/geos/factory.py [hunk 2]
```diff
@@ -15,12 +17,10 @@ def fromfile(file_h):
 
     # If we get WKB need to wrap in memoryview(), so run through regexes.
     if isinstance(buf, bytes):
-        try:
+        with suppress(UnicodeDecodeError):
             decoded = buf.decode()
```
Comment (timgraham) [inline django/contrib/gis/geos/factory.py]: Not certain, but it seems like maybe only this line should be in the context manager.

File: django/contrib/gis/utils/__init__.py [hunk 3]
```diff
@@ -1,16 +1,16 @@
...
 from django.core.exceptions import ImproperlyConfigured
 
 if HAS_GDAL:
     from django.contrib.gis.utils.ogrinfo import ogrinfo  # NOQA
     from django.contrib.gis.utils.ogrinspect import mapping, ogrinspect  # NOQA
     from django.contrib.gis.utils.srs import add_srs_entry  # NOQA
-    try:
+    with suppress(ImproperlyConfigured):
         # LayerMapping requires DJANGO_SETTINGS_MODULE to be set,
         # so this needs to be in try/except.
```
Comment (timgraham) [inline django/contrib/gis/utils/__init__.py]: update comment: "ImproperlyConfigured is raised if that's not the case."

File: django/core/files/move.py [hunk 4]
```diff
@@ -41,17 +42,16 @@ def file_move_safe(old_file_name, new_file_name, chunk_size=1024 * 64, allow_ove
     if _samefile(old_file_name, new_file_name):
         return
 
-    try:
+    with suppress(OSError):
+        # This will happen with os.rename if moving to another filesystem
```
Comment (timgraham) [inline django/core/files/move.py]: OSError will happen... os.rename()

File: django/core/files/move.py [hunk 5]
```diff
@@ -41,17 +42,16 @@ def file_move_safe(old_file_name, new_file_name, chunk_size=1024 * 64, allow_ove
...
+        # or when moving opened files on certain operating systems
```
Comment (timgraham) [inline django/core/files/move.py]: add period

File: django/core/validators.py [hunk 6]
```diff
@@ -201,12 +202,10 @@ def __call__(self, value):
         if (domain_part not in self.domain_whitelist and
                 not self.validate_domain_part(domain_part)):
             # Try for possible IDN domain-part
-            try:
+            with suppress(UnicodeError):
                 domain_part = domain_part.encode('idna').decode('ascii')
                 if self.validate_domain_part(domain_part):
```
Comment (timgraham) [inline django/core/validators.py]: Not sure this should be in the context manager. Technically it would probably be in the else of try/except/else.

File: django/db/backends/oracle/introspection.py [hunk 7]
```diff
@@ -22,15 +23,11 @@ class DatabaseIntrospection(BaseDatabaseIntrospection):
         cx_Oracle.TIMESTAMP: 'DateTimeField',
     }
 
-    try:
+    with suppress(AttributeError):
         data_types_reverse[cx_Oracle.NATIVE_FLOAT] = 'FloatField'
```
Comment (timgraham) [inline django/db/backends/oracle/introspection.py]: I guess the `AttributeError` is for `cx_Oracle.NATIVE_FLOAT` -- this is probaby obsolete now that cx Oracle 5.2+ is supported.

File: django/db/backends/postgresql/client.py [hunk 8]
```diff
@@ -38,7 +39,9 @@ def runshell_db(cls, conn_params):
             if passwd:
                 # Create temporary .pgpass file.
                 temp_pgpass = NamedTemporaryFile(mode='w+')
-                try:
+                with suppress(UnicodeEncodeError):
+                    # If the current locale can't encode the data, we let
```
Comment (timgraham) [inline django/db/backends/postgresql/client.py]: chop "we"

File: django/db/backends/sqlite3/operations.py [hunk 9]
```diff
@@ -29,18 +30,16 @@ def check_expression_support(self, expression):
         bad_aggregates = (aggregates.Sum, aggregates.Avg, aggregates.Variance, aggregates.StdDev)
         if isinstance(expression, bad_aggregates):
             for expr in expression.get_source_expressions():
-                try:
+                with suppress(FieldError):
+                    # Not every subexpression has an output_field which is fine
+                    # to ignore.
                     output_field = expr.output_field
                     if isinstance(output_field, bad_fields):
```
Comment (timgraham) [inline django/db/backends/sqlite3/operations.py]: Here again the try block isn't minimized.

File: django/db/models/query.py [hunk 10]
```diff
@@ -1263,12 +1262,10 @@ def columns(self):
         columns = self.query.get_columns()
         # Adjust any column names which don't match field names
         for (query_name, model_name) in self.translations.items():
-            try:
+            with suppress(ValueError):
+                # Ignore translations for nonexistent column names
```
Comment (timgraham) [inline django/db/models/query.py]: It feels more natural to me to put these type of comments before with statement rather than after it. Do you have any feelings about that?

Comment (atombrella) [inline django/db/models/query.py]: Not really. I considered this when finding the `try`-blocks, and where to put the comment, and ended up doing it that way. I'll move them one line up :)

File: django/db/backends/oracle/introspection.py (hunk 7 above)
Comment (atombrella) [inline django/db/backends/oracle/introspection.py]: Good. I've moved the data type into the map instead. There's a remark about `Unicode` in http://cx-oracle.readthedocs.io/en/latest/releasenotes.html?highlight=unicode (I won't change anything here, but probably those data types should be in the reverse map as well)?

Comment (timgraham) [inline django/db/backends/oracle/introspection.py]: Let's revert this file and have an Oracle expert handle the changes in https://code.djangoproject.com/ticket/27822#ticket.

File: django/contrib/gis/gdal/srs.py [hunk 11]
```diff
@@ -55,14 +56,11 @@ def __init__(self, srs_input='', srs_type='user'):
...
         elif isinstance(srs_input, str):
-            try:
+            with suppress(ValueError):
                 # If SRID is a string, e.g., '4326', then make acceptable
                 # as user input.
                 srid = int(srs_input)
-                srs_input = 'EPSG:%d' % srid
-            except ValueError:
-                pass
-        elif isinstance(srs_input, int):
```
Comment (timgraham) [inline django/contrib/gis/gdal/srs.py]: Looks like the test failures are because this line was removed.

File: django/contrib/gis/geos/factory.py (hunk 2 above)
Comment (timgraham) [inline django/contrib/gis/geos/factory.py]: Now I see that my comments to this effect are actually changing the semantics slightly -- previously the if statement would be skipped if `UnicodeDecodeError` happens, now they'll be run. I'm not confident if these changes are correct so I guess it's better to leave things as they were. I  wonder if try/except/else would be better than `suppress()`.

Comment (atombrella) [inline django/contrib/gis/geos/factory.py]: I've reverted this change, and `srs.py` as well. We can close the ticket as wontfix, possible. The effort was not completely in vain since it spawned a new ticket about the Oracle datatypes.

File: django/core/files/move.py [hunk 12]
```diff
@@ -41,17 +42,15 @@ def file_move_safe(old_file_name, new_file_name, chunk_size=1024 * 64, allow_ove
     if _samefile(old_file_name, new_file_name):
         return
 
-    try:
-        # If the destination file exists and allow_overwrite is False then raise an IOError
+    # OSError will happen with os.rename() if moving to another filesystem
+    # or when moving opened files on certain operating systems.
+    with suppress(OSError):
+        # If the destination file exists and allow_overwrite is False then raise an IOError.
```
Comment (timgraham) [inline django/core/files/move.py]: This comment can be removed as it's repeating what the error message says.

File: django/template/backends/base.py [hunk 13]
```diff
@@ -73,9 +75,7 @@ def iter_template_filenames(self, template_name):
         directory traversal attacks.
         """
         for template_dir in self.template_dirs:
-            try:
-                yield safe_join(template_dir, template_name)
-            except SuspiciousFileOperation:
+            with suppress(SuspiciousFileOperation):
                 # The joined path was located outside of this template_dir
```
Comment (timgraham) [inline django/template/backends/base.py]: There are still a few cases like this where the comment's connection to `suppress()` is less obvious than it was when try/except was used. You could move this above the with and rephrase it like "SuspiciousFileOperation occurs if the jointed path is located outside...."

File: django/contrib/gis/db/models/fields.py [hunk 14]
```diff
@@ -435,13 +434,11 @@ def contribute_to_class(self, cls, name, **kwargs):
         setattr(cls, self.attname, SpatialProxy(gdal.GDALRaster, self))
 
     def get_transform(self, name):
-        try:
+        with suppress(ValueError):
             band_index = int(name)
```
Comment (timgraham) [inline django/contrib/gis/db/models/fields.py]: Overall, the changes are mostly fine, I just wonder if we might use this as an opportunity to tighten up large try blocks and use try/except/else instead (such as here).

File: tests/transactions/tests.py [hunk 15]
```diff
@@ -220,14 +221,12 @@ def test_prevent_rollback(self):
 
     @skipIf(sys.platform.startswith('win'), "Windows doesn't have signals.")
     def test_rollback_on_keyboardinterrupt(self):
-        try:
+        with suppress(KeyboardInterrupt):
```
Comment (pope1ni) [inline tests/transactions/tests.py]: Can this be combined with the line below?

```python
with suppress(KeyboardInterrupt), transaction.atomic():
```

While I've highlighted this one, there are a lot more places where this could be done. It would remove excess levels of indentation, but may be less clear in certain cases, e.g. nested `transaction.atomic()` for savepoints.

I'm not massively pushing for this, just thought I'd put it out there.

**Comment by mfbx9da4** [comment]
Hi @atombrella I would like to help out on this PR. Would it be possible to get write permissions to your fork so that I can add some commits here?

**Comment by atombrella** [comment]
@mfbx9da4 Could you elaborate a bit what you have in mind? I'd prefer a PR against the branch in my local fork. Thanks.

**Comment by mfbx9da4** [comment]
@atombrella I tried creating a PR against your local fork but github couldn't find / doesn't allow that for some reason
image

I simply want to do as @pope1ni suggested and update any cases you may have missed 😄

**Comment by pope1ni** [comment]
So when I commented about merging the with statements together, I implied that there were many other areas where this could / should be done, not just that one. I was only suggesting to perhaps be conservative in certain cases where it may be confusing, e.g.

```python
with transaction.atomic(), transaction.atomic():
```

Having said that, it should be expected that the first is either a transaction or savepoint and the second can only be a savepoint. Such a construct would be unusual anyway - and likely completely pointless - it probably only happens in the tests...

**Comment by atombrella** [comment]
@pope1ni @mfbx9da4 a `grep` finds statements in `tests/mail` and `tests/annotations` that could be merged. There are only a few of these cases with multiple with-statements. Locally, the test completes successfully when I move `backend` to the outer `with`-statement in the mail testsuite.

There's one or two try-except-pass cases left that I reverted back, where the comment is quite long.

File: django/contrib/gis/db/models/fields.py [hunk 16]
```diff
@@ -444,4 +443,5 @@ def get_transform(self, name):
             )
         except ValueError:
             pass
-        return super().get_transform(name)
+        else:
```
Comment (timgraham) [inline django/contrib/gis/db/models/fields.py]: Why is this changed? I think replacing `pass` with `return None` would be more explicit if that's really the correct logic.

File: tests/cache/tests.py [hunk 17]
```diff
@@ -1223,6 +1223,7 @@ def test_memcached_deletes_key_on_failed_set(self):
             # was deleted, so the return/exception behavior for the set()
             # itself is not important.
             pass
+
```
Comment (timgraham) [inline tests/cache/tests.py]: revert

**Comment by mfbx9da4** [comment]
@pope1ni I intend to find and fix other cases. I just made an initial commit to test the waters to see if I could contribute here 😄 

@atombrella I see you rebased my commit into your commit to keep things tidy. To be honest, I would quite like to have a commit in the final PR so I can officially become a contributor. If I separated out some functionality ie joining `with ` where necessary throughout in my commit would you consider leaving it as a separate commit?

**Comment by timgraham** [comment]
David, unless your changes are remodifying places that this PR changes, they are better done in a separate PR anyway.

**Comment by mfbx9da4** [comment]
Okay no worries

File: django/contrib/gis/db/models/fields.py (hunk 16 above)
Comment (timgraham) [inline django/contrib/gis/db/models/fields.py]: Is some test failing without this change? I don't see how it's related to this PR.

**Comment by atombrella** [comment]
https://dpaste.de/L0bA/raw coverage changes a little bit, mostly +/-1 percent point.
